README.md
# @sumwatshade/oclif-plugin-update

====================

> A fork of @oclif/plugin-update that closer mimics [NVM](https://github.com/nvm-sh/nvm)

[![Version](https://img.shields.io/npm/v/@oclif/plugin-update.svg)](https://npmjs.org/package/@sumwatshade/oclif-plugin-update)
[![CircleCI](https://circleci.com/gh/oclif/plugin-update/tree/master.svg?style=shield)](https://circleci.com/gh/sumwatshade/plugin-update/tree/master)
[![Downloads/week](https://img.shields.io/npm/dw/@oclif/plugin-update.svg)](https://npmjs.org/package/@sumwatshade/oclif-plugin-update)
[![License](https://img.shields.io/npm/l/@oclif/plugin-update.svg)](https://github.com/sumwatshade/plugin-update/blob/master/package.json)

<!-- toc -->
* [@sumwatshade/oclif-plugin-update](#sumwatshadeoclif-plugin-update)
<!-- tocstop -->

## Commands

<!-- commands -->
* [`oclif-example install VERSION`](#oclif-example-install-version)
* [`oclif-example update [CHANNEL]`](#oclif-example-update-channel)
* [`oclif-example use [VERSION]`](#oclif-example-use-version)

## `oclif-example install VERSION`

Install and link a new version of the oclif-example CLI. This will first check locally before fetching from the internet

```
USAGE
  $ oclif-example install VERSION

ARGUMENTS
  VERSION  Specify an explicit version (ex. 3.0.0-next.1) or a channel (ex. alpha)
```

_See code: [src/commands/install.ts](https://github.com/sumwatshade/plugin-update/blob/v1.9.4/src/commands/install.ts)_

## `oclif-example update [CHANNEL]`

Updates the oclif-example CLI. This will check for the latest version available on the requested channel and fetch it from remote

```
USAGE
  $ oclif-example update [CHANNEL]

ARGUMENTS
  CHANNEL  Specify a channel (ex: stable,alpha,beta,next). An error will be thrown if this channel is invalid.

OPTIONS
  --from-local  interactively choose an already installed version
```

_See code: [src/commands/update.ts](https://github.com/sumwatshade/plugin-update/blob/v1.9.4/src/commands/update.ts)_

## `oclif-example use [VERSION]`

Checks for a previously installed version of the oclif-example CLI. Throws an error if the version is not found.

```
USAGE
  $ oclif-example use [VERSION]

ARGUMENTS
  VERSION  Specify an explicit version (ex. 3.0.0-next.1) or a channel (ex. alpha)
```

_See code: [src/commands/use.ts](https://github.com/sumwatshade/plugin-update/blob/v1.9.4/src/commands/use.ts)_
<!-- commandsstop -->


CHANGELOG.md
## [1.9.4](https://github.com/sumwatshade/plugin-update/compare/v1.9.3...v1.9.4) (2021-09-27)


### Bug Fixes

* use 14 for publish ([a74451d](https://github.com/sumwatshade/plugin-update/commit/a74451d67417ad64a52578f0128c14b3fcb605a6))
* work in node 10 ([#38](https://github.com/sumwatshade/plugin-update/issues/38)) ([1b01541](https://github.com/sumwatshade/plugin-update/commit/1b01541d0464b0d37f694f2b1bb98610c651ab5e))

## [1.9.3](https://github.com/sumwatshade/plugin-update/compare/v1.9.2...v1.9.3) (2021-09-27)


### Bug Fixes

* deps and build ([b6cbc15](https://github.com/sumwatshade/plugin-update/commit/b6cbc158b12e72c883d88a6598ee66af8557fe40))
* node 10 in prs ([a6e1094](https://github.com/sumwatshade/plugin-update/commit/a6e1094fc5d190ec2dc4b98458344304363d139d))
* use github actions ([7a92029](https://github.com/sumwatshade/plugin-update/commit/7a920299e9f5387a1dca562dde50ea071b0a58df))

## [1.9.2](https://github.com/sumwatshade/plugin-update/compare/v1.9.1...v1.9.2) (2021-08-26)


### Bug Fixes

* null checking on client root ([fed04a3](https://github.com/sumwatshade/plugin-update/commit/fed04a39a58f86c0ee69b8626b57b67f61c1816b))

## [1.9.1](https://github.com/sumwatshade/plugin-update/compare/v1.9.0...v1.9.1) (2021-08-23)


### Bug Fixes

* code organization and prep for requests around installation aliases ([1c7791d](https://github.com/sumwatshade/plugin-update/commit/1c7791d65823b87c1b3e6ae10099a403f0aab9b6))

# [1.9.0](https://github.com/sumwatshade/plugin-update/compare/v1.8.6...v1.9.0) (2021-08-04)


### Bug Fixes

* small tweak ([70e7674](https://github.com/sumwatshade/plugin-update/commit/70e76746568674fc343e47dcf275ac9cdc5c1f48))


### Features

* provide major versions ([9ac3c71](https://github.com/sumwatshade/plugin-update/commit/9ac3c7172a57168899e05fca802ff54599fa031a))

## [1.8.6](https://github.com/sumwatshade/plugin-update/compare/v1.8.5...v1.8.6) (2021-06-24)


### Bug Fixes

* extract symlinks from node modules ([52fc421](https://github.com/sumwatshade/plugin-update/commit/52fc4212fda8475bc1b0141720514af7dcc373f3))
* symlink resolution for older builds ([fbd48df](https://github.com/sumwatshade/plugin-update/commit/fbd48dff5ff6be52b1df7bb1858855d2205f88b5))

## [1.8.5](https://github.com/sumwatshade/plugin-update/compare/v1.8.4...v1.8.5) (2021-06-23)


### Bug Fixes

* add docs for each command ([1066934](https://github.com/sumwatshade/plugin-update/commit/1066934dd31bb0eb2e4371a557184e5f5054bd9d))

## [1.8.4](https://github.com/sumwatshade/plugin-update/compare/v1.8.3...v1.8.4) (2021-06-17)


### Bug Fixes

* no reexec on install ([f55aa79](https://github.com/sumwatshade/plugin-update/commit/f55aa795905f89eacc2579556b8cdb196bdc6aa0))

## [1.8.3](https://github.com/sumwatshade/plugin-update/compare/v1.8.2...v1.8.3) (2021-06-17)


### Bug Fixes

* check if channelUpdateRequested is req ([4685a72](https://github.com/sumwatshade/plugin-update/commit/4685a72dfa8a7b37075a727990fe53d2ada8cd44))
* remove sha when extracting target version ([67dfe55](https://github.com/sumwatshade/plugin-update/commit/67dfe553a4a728a6d9d988aa69c94ae7b321afcf))
* update comment ([2f8dc5c](https://github.com/sumwatshade/plugin-update/commit/2f8dc5cbb0b91b18a84d05eab50bb42631a9414a))

## [1.8.2](https://github.com/sumwatshade/plugin-update/compare/v1.8.1...v1.8.2) (2021-06-17)


### Bug Fixes

* udpate channel to stable if version request is not a diff channel ([5314226](https://github.com/sumwatshade/plugin-update/commit/531422650d2d7a3aa5dc0ecbcd2d97c6837badc3))

## [1.8.1](https://github.com/sumwatshade/plugin-update/compare/v1.8.0...v1.8.1) (2021-06-16)


### Bug Fixes

* add success msg + fix tar ext ([e7c5729](https://github.com/sumwatshade/plugin-update/commit/e7c5729fe0fae1382b5f1894155d058c9518b644))

# [1.8.0](https://github.com/sumwatshade/plugin-update/compare/v1.7.5...v1.8.0) (2021-06-15)


### Bug Fixes

* add explicit version check ([6a41c1a](https://github.com/sumwatshade/plugin-update/commit/6a41c1a65d5b39c919cc23c7dcc18212f53d483c))
* errors ([6e0eb3d](https://github.com/sumwatshade/plugin-update/commit/6e0eb3dbfe487d8b2cfc5d97c6efd353c28c8181))
* lint ([3be6770](https://github.com/sumwatshade/plugin-update/commit/3be6770c113e89e1b4ff3b282f68a7e5a7259137))
* tests ([eea1b80](https://github.com/sumwatshade/plugin-update/commit/eea1b80a8644e2c229bc03f800f7b63fa5f47e11))
* tests ([d0a8e3e](https://github.com/sumwatshade/plugin-update/commit/d0a8e3e587b4d2f6c1f6aa24019276589e939034))
* use config ([01c6f9d](https://github.com/sumwatshade/plugin-update/commit/01c6f9d8705a4cd5bc7044c85a2f70f08d58c2ba))


### Features

* pass in specific version ([7bbb48e](https://github.com/sumwatshade/plugin-update/commit/7bbb48e942e7d5b3fdb351f42832acc242c9b1f6))
* update install command ([6cecef6](https://github.com/sumwatshade/plugin-update/commit/6cecef6c7d870ec6c6aa7d215b5063b9dd9b8a0c))

## [1.7.5](https://github.com/sumwatshade/plugin-update/compare/v1.7.4...v1.7.5) (2021-06-04)


### Bug Fixes

* configuration changes ([8638fbd](https://github.com/sumwatshade/plugin-update/commit/8638fbd771842317e5d6b8a73af604bd691d0de4))
* first pass formatting ([5d82adc](https://github.com/sumwatshade/plugin-update/commit/5d82adca2cff7fade36f7832e17fec6640eb11f8))
* prettier ([142ef3e](https://github.com/sumwatshade/plugin-update/commit/142ef3ec0f9a1f7b24ae072b1b8cf35579877521))
* respect prettier config in ide ([2382a3e](https://github.com/sumwatshade/plugin-update/commit/2382a3ef9906b5f3d5571992b5f4ec1babc8de28))
* run lint on CI ([0e288a7](https://github.com/sumwatshade/plugin-update/commit/0e288a7dae5f0f463b2254df6828045e9149c743))
* update formatting ([ab71f47](https://github.com/sumwatshade/plugin-update/commit/ab71f47117d0cea413480a1b0236e513e75ca6d1))

## [1.7.4](https://github.com/sumwatshade/plugin-update/compare/v1.7.3...v1.7.4) (2021-06-03)


### Bug Fixes

* strip partials ([56383cf](https://github.com/sumwatshade/plugin-update/commit/56383cfaf85a731fc712e67fd33305a41d8ee153))


src/commands/update.ts
import color from '@oclif/color';
import Command, { flags } from '@oclif/command';
import { IManifest } from '@oclif/dev-cli';
import cli from 'cli-ux';
import * as spawn from 'cross-spawn';
import * as fs from 'fs-extra';
import HTTP from 'http-call';
import * as _ from 'lodash';
import * as path from 'path';

import { extract } from '../tar';
import { ls, wait } from '../util';
import { isSemverAlias } from '../utils/version-calculation';

export default class UpdateCommand extends Command {
  static description =
    'Updates the <%= config.bin %> CLI. This will check for the latest version available on the requested channel and fetch it from remote';

  static args = [
    {
      name: 'channel', // name of arg to show in help and reference with args[name]
      required: false, // make the arg required with `required: true`
      description:
        'Specify a channel (ex: stable,alpha,beta,next). An error will be thrown if this channel is invalid.', // help description
    },
  ];

  static flags: flags.Input<any> = {
    autoupdate: flags.boolean({ hidden: true }),
    'from-local': flags.boolean({
      description: 'interactively choose an already installed version',
    }),
  };

  protected autoupdate!: boolean;

  protected channel!: string;

  protected currentVersion?: string;

  protected updatedVersion!: string;

  protected readonly clientRoot =
    this.config.scopedEnvVar('OCLIF_CLIENT_HOME') ||
    path.join(this.config.dataDir, 'client');

  protected readonly clientBin = path.join(
    this.clientRoot,
    'bin',
    this.config.windows ? `${this.config.bin}.cmd` : this.config.bin,
  );

  async run() {
    const { args, flags } = this.parse(UpdateCommand);
    this.autoupdate = Boolean(flags.autoupdate);

    if (this.autoupdate) await this.debounce();

    this.channel = args.channel || (await this.determineChannel());

    if (await isSemverAlias(this.channel)) {
      throw new Error(
        `We do not yet support major/minor aliases with the update command. Please refer to this issue for updates on the enhancement: https://github.com/sumwatshade/plugin-update/issues/32`,
      );
    }

    if (flags['from-local']) {
      await this.ensureClientDir();
      this.debug(
        `Looking for locally installed versions at ${this.clientRoot}`,
      );

      // Do not show known non-local version folder names, bin and current.
      let versions: string[] = [];
      try {
        versions = fs
          .readdirSync(this.clientRoot)
          .filter(
            (dirOrFile) => dirOrFile !== 'bin' && dirOrFile !== 'current',
          );
      } catch {}

      if (versions.length === 0)
        throw new Error('No locally installed versions found.');

      this.log(
        `Found versions: \n${versions
          .map((version) => `     ${version}`)
          .join('\n')}\n`,
      );

      const pinToVersion = await cli.prompt('Enter a version to update to');
      if (!versions.includes(pinToVersion))
        throw new Error(
          `Version ${pinToVersion} not found in the locally installed versions.`,
        );

      if (!(await fs.pathExists(path.join(this.clientRoot, pinToVersion)))) {
        throw new Error(
          `Version ${pinToVersion} is not already installed at ${this.clientRoot}.`,
        );
      }
      cli.action.start(`${this.config.name}: Updating CLI`);
      this.debug(`switching to existing version ${pinToVersion}`);
      this.updateToExistingVersion(pinToVersion);

      this.log();
      this.log(
        `Updating to an already installed version will not update the channel. If autoupdate is enabled, the CLI will eventually be updated back to ${this.channel}.`,
      );
    } else {
      cli.action.start(`${this.config.name}: Updating CLI`);
      await this.config.runHook('preupdate', { channel: this.channel });
      const manifest = await this.fetchManifest();
      this.currentVersion = await this.determineCurrentVersion();
      this.updatedVersion = (manifest as any).sha
        ? `${manifest.version}-${(manifest as any).sha}`
        : manifest.version;
      const reason = await this.skipUpdate();
      if (reason) cli.action.stop(reason || 'done');
      else await this.update(manifest);
      this.debug('tidy');
      await this.tidy();
      await this.config.runHook('update', { channel: this.channel });
    }

    this.debug('done');
    cli.action.stop();
  }

  protected async fetchManifest(): Promise<IManifest> {
    const http: typeof HTTP = require('http-call').HTTP;

    cli.action.status = 'fetching manifest';
    if (!this.config.scopedEnvVarTrue('USE_LEGACY_UPDATE')) {
      try {
        const newManifestUrl = this.config.s3Url(
          this.s3ChannelManifestKey(
            this.config.bin,
            this.config.platform,
            this.config.arch,
            (this.config.pjson.oclif.update.s3 as any).folder,
          ),
        );
        const { body } = await http.get<IManifest | string>(newManifestUrl);
        if (typeof body === 'string') {
          return JSON.parse(body);
        }
        return body;
      } catch (error) {
        this.debug(error.message);
      }
    }

    try {
      const url = this.config.s3Url(
        this.config.s3Key('manifest', {
          channel: this.channel,
          platform: this.config.platform,
          arch: this.config.arch,
        }),
      );
      const { body } = await http.get<IManifest | string>(url);

      // in case the content-type is not set, parse as a string
      // this will happen if uploading without `oclif-dev publish`
      if (typeof body === 'string') {
        return JSON.parse(body);
      }
      return body;
    } catch (error) {
      if (error.statusCode === 403)
        throw new Error(
          `HTTP 403: This could mean that you are being restricted from accessing this version, or that you are providing an invalid channel (${this.channel})`,
        );
      throw error;
    }
  }

  protected async downloadAndExtract(
    output: string,
    manifest: IManifest,
    channel: string,
    targetVersion?: string,
  ) {
    const { version } = manifest;

    const filesize = (n: number): string => {
      const [num, suffix] = require('filesize')(n, { output: 'array' });
      return num.toFixed(1) + ` ${suffix}`;
    };

    const http: typeof HTTP = require('http-call').HTTP;
    const gzUrl =
      !targetVersion && manifest.gz
        ? manifest.gz
        : this.config.s3Url(
            this.config.s3Key('versioned', {
              version: targetVersion,
              channel,
              bin: this.config.bin,
              platform: this.config.platform,
              arch: this.config.arch,
              ext: targetVersion ? '.tar.gz' : 'gz',
            }),
          );
    const { response: stream } = await http.stream(gzUrl);
    stream.pause();

    const baseDir =
      manifest.baseDir ||
      this.config.s3Key('baseDir', {
        version,
        channel,
        bin: this.config.bin,
        platform: this.config.platform,
        arch: this.config.arch,
      });
    const extraction = extract(
      stream,
      baseDir,
      output,
      !targetVersion && manifest.sha256gz ? manifest.sha256gz : undefined,
    );

    // to-do: use cli.action.type
    if ((cli.action as any).frames) {
      // if spinner action
      const total = parseInt(stream.headers['content-length']!, 10);
      let current = 0;
      const updateStatus = _.throttle(
        (newStatus: string) => {
          cli.action.status = newStatus;
        },
        250,
        { leading: true, trailing: false },
      );
      stream.on('data', (data) => {
        current += data.length;
        updateStatus(`${filesize(current)}/${filesize(total)}`);
      });
    }

    stream.resume();
    await extraction;
  }

  protected async update(
    manifest: IManifest,
    channel = this.channel,
    targetVersion?: string,
    reexec = true,
  ) {
    const { channel: manifestChannel } = manifest;
    if (manifestChannel) channel = manifestChannel;
    cli.action.start(
      `${this.config.name}: Updating CLI from ${color.green(
        this.currentVersion,
      )} to ${color.green(this.updatedVersion)}${
        channel === 'stable' ? '' : ' (' + color.yellow(channel) + ')'
      }`,
    );

    await this.ensureClientDir();
    const output = path.join(this.clientRoot, this.updatedVersion);

    if (!(await fs.pathExists(output))) {
      await this.downloadAndExtract(output, manifest, channel, targetVersion);
    }

    await this.setChannel();
    await this.createBin(this.updatedVersion);
    await this.touch();
    if (reexec) {
      await this.reexec();
    }
  }

  protected async updateToExistingVersion(version: string) {
    await this.createBin(version);
    await this.touch();
  }

  protected async skipUpdate(): Promise<string | false> {
    if (!this.config.binPath) {
      const instructions = this.config.scopedEnvVar('UPDATE_INSTRUCTIONS');
      if (instructions) this.warn(instructions);
      return 'not updatable';
    }
    if (this.currentVersion === this.updatedVersion) {
      if (this.config.scopedEnvVar('HIDE_UPDATED_MESSAGE')) return 'done';
      return `already on latest version: ${this.currentVersion}`;
    }
    return false;
  }

  protected async determineChannel(): Promise<string> {
    const channelPath = path.join(this.config.dataDir, 'channel');
    this.debug(`Reading channel from ${channelPath}`);
    if (fs.existsSync(channelPath)) {
      const channel = await fs.readFile(channelPath, 'utf8');
      this.debug(`Read channel from data: ${String(channel)}`);
      return String(channel).trim();
    }
    return this.config.channel || 'stable';
  }

  protected async determineCurrentVersion(): Promise<string | undefined> {
    try {
      const currentVersion = await fs.readFile(this.clientBin, 'utf8');
      const matches = currentVersion.match(/\.\.[/|\\](.+)[/|\\]bin/);
      return matches ? matches[1] : this.config.version;
    } catch (error) {
      this.debug(error);
    }
    return this.config.version;
  }

  protected s3ChannelManifestKey(
    bin: string,
    platform: string,
    arch: string,
    folder?: string,
  ): string {
    let s3SubDir = folder || '';
    if (s3SubDir !== '' && s3SubDir.slice(-1) !== '/')
      s3SubDir = `${s3SubDir}/`;
    return path.join(
      s3SubDir,
      'channels',
      this.channel,
      `${bin}-${platform}-${arch}-buildmanifest`,
    );
  }

  protected async setChannel() {
    const channelPath = path.join(this.config.dataDir, 'channel');
    this.debug(`Writing channel (${this.channel}) to ${channelPath}`);
    fs.writeFile(channelPath, this.channel, 'utf8');
  }

  protected async logChop() {
    try {
      this.debug('log chop');
      const logChopper = require('log-chopper').default;
      await logChopper.chop(this.config.errlog);
    } catch (error) {
      this.debug(error.message);
    }
  }

  protected async mtime(f: string) {
    const { mtime } = await fs.stat(f);
    return mtime;
  }

  // when autoupdating, wait until the CLI isn't active
  protected async debounce(): Promise<void> {
    let output = false;
    const lastrunfile = path.join(this.config.cacheDir, 'lastrun');
    const m = await this.mtime(lastrunfile);
    m.setHours(m.getHours() + 1);
    if (m > new Date()) {
      const msg = `waiting until ${m.toISOString()} to update`;
      if (output) {
        this.debug(msg);
      } else {
        await cli.log(msg);
        output = true;
      }
      await wait(60 * 1000); // wait 1 minute
      return this.debounce();
    }
    cli.log('time to update');
  }

  // removes any unused CLIs
  protected async tidy() {
    try {
      const root = this.clientRoot;
      if (!(await fs.pathExists(root))) return;
      const files = await ls(root);
      const promises = files.map(async (f) => {
        if (
          ['bin', 'current', this.config.version].includes(
            path.basename(f.path),
          )
        )
          return;
        const mtime = f.stat.mtime;
        mtime.setHours(mtime.getHours() + 14 * 24);
        if (mtime < new Date()) {
          await fs.remove(f.path);
        }
      });
      for (const p of promises) await p; // eslint-disable-line no-await-in-loop
      await this.logChop();
    } catch (error) {
      cli.warn(error);
    }
  }

  protected async touch() {
    // touch the client so it won't be tidied up right away
    try {
      const p = path.join(this.clientRoot, this.config.version);
      this.debug('touching client at', p);
      if (!(await fs.pathExists(p))) return;
      await fs.utimes(p, new Date(), new Date());
    } catch (error) {
      this.warn(error);
    }
  }

  protected async reexec() {
    cli.action.stop();
    return new Promise((_, reject) => {
      this.debug('restarting CLI after update', this.clientBin);
      const commandArgs = ['update', this.channel ? this.channel : ''].filter(
        Boolean,
      );
      spawn(this.clientBin, commandArgs, {
        stdio: 'inherit',
        env: {
          ...process.env,
          [this.config.scopedEnvVarKey('HIDE_UPDATED_MESSAGE')]: '1',
        },
      })
        .on('error', reject)
        .on('close', (status: number) => {
          try {
            if (status > 0) this.exit(status);
          } catch (error) {
            reject(error);
          }
        });
    });
  }

  protected async createBin(version: string) {
    const dst = this.clientBin;
    const { bin } = this.config;
    const binPathEnvVar = this.config.scopedEnvVarKey('BINPATH');
    const redirectedEnvVar = this.config.scopedEnvVarKey('REDIRECTED');
    if (this.config.windows) {
      const body = `@echo off
setlocal enableextensions
set ${redirectedEnvVar}=1
set ${binPathEnvVar}=%~dp0${bin}
"%~dp0..\\${version}\\bin\\${bin}.cmd" %*
`;
      await fs.outputFile(dst, body);
    } else {
      /* eslint-disable no-useless-escape */
      const body = `#!/usr/bin/env bash
set -e
get_script_dir () {
  SOURCE="\${BASH_SOURCE[0]}"
  # While $SOURCE is a symlink, resolve it
  while [ -h "$SOURCE" ]; do
    DIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
    SOURCE="$( readlink "$SOURCE" )"
    # If $SOURCE was a relative symlink (so no "/" as prefix, need to resolve it relative to the symlink base directory
    [[ $SOURCE != /* ]] && SOURCE="$DIR/$SOURCE"
  done
  DIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
  echo "$DIR"
}
DIR=$(get_script_dir)
${binPathEnvVar}="\$DIR/${bin}" ${redirectedEnvVar}=1 "$DIR/../${version}/bin/${bin}" "$@"
`;
      /* eslint-enable no-useless-escape */

      await fs.remove(dst);
      await fs.outputFile(dst, body);
      await fs.chmod(dst, 0o755);
      await fs.remove(path.join(this.clientRoot, 'current'));
      await fs.symlink(`./${version}`, path.join(this.clientRoot, 'current'));
    }
  }

  protected async ensureClientDir() {
    try {
      await fs.mkdirp(this.clientRoot);
    } catch (error) {
      if (error.code === 'EEXIST') {
        // for some reason the client directory is sometimes a file
        // if so, this happens. Delete it and recreate
        await fs.remove(this.clientRoot);
        await fs.mkdirp(this.clientRoot);
      } else {
        throw error;
      }
    }
  }
}
